### ISSUE TITLE
Failing to read "LastPlayTime"

### ISSUE THREAD

**Issue by FrogTheFrog** [tirish/steam-shortcut-editor#7]
Steam add `LastPlayTime` in some sort of raw value in milliseconds, I guess and parser tries to read it as boolean and it fails. Providing the file I'm testing it on.

[shortcuts4.zip](https://github.com/tirish/steam-shortcut-editor/files/968140/shortcuts4.zip)

**Comment by FrogTheFrog** [comment]
I have managed to make a temporary fix, but I have no idea how to fix it properly. I'll be using my fork until you fix it. Also, please do not use `console.log | console.error` to handle errors, leave it to us.

[My fork.](https://github.com/FrogTheFrog/steam-shortcut-editor.git)

**Comment by tirish** [comment]
Thanks for flagging this. I'll try and find some time this week to look into it.

**Comment by tirish** [comment]
The characters immediately following `LastPlayTime` are:
```
  : 0
� : 65533
6 : 54
 : 7
Y : 89
  : 0
```
(Left of `:` being utf8 character, right being character code)
Snippet: `\u0002LastPlayTime\u0000�6\u0007Y\u0000tags`

@FrogTheFrog do you know what the actual value is supposed to be for this property?

**Comment by SkyNinja** [comment]
They should relate to the last time that particular game was played. I have two games which report the same time in Steam's UI, but are different in shortcuts.vdf.

One is `\u0002LastPlayTime\u0000M\u0012°X\u0000tags`
The other is `\u0002LastPlayTime\u0000h×¯X\u0000tags`
Both of these display as 2/24/2017 in the client. I did notice that the last two sets characters are very close, with the degree symbol being character 176 and the macron 175.

Also, the default, "never played" state is just filled with nul characters.

**Comment by tirish** [comment]
I believe the four characters that follow `\u0002LastPlayTime\u0000` should be treated as binary. Not sure if the value should be 32bits or 64bits (but it is likely 32). Not sure how to go from a single binary number to a date though. Plugging the 32bit integer into `new Date(...)` does not result in the right date, and the two values provided by @SkyNinja result in wildly different binary representations.

```
M°X
M : 77 : 01001101 : 0000000001001101
 : 18 : 00010010 : 0000000000010010
° : 176 : 10110000 : 0000000010110000
X : 88 : 01011000 : 0000000001011000
Character Codes | Sum: 359 ; Concat: 771817688
Binary Int32 |  01001101000100101011000001011000 ; Int: 1293070424 ; Date (Int * 1000): 2010-12-23T02:13:44.000Z
Binary Int64 |  0000000001001101000000000001001000000000101100000000000001011000


h×¯X
h : 104 : 01101000 : 0000000001101000
× : 215 : 11010111 : 0000000011010111
¯ : 175 : 10101111 : 0000000010101111
X : 88 : 01011000 : 0000000001011000
Character Codes | Sum: 582 ; Concat: 10421517588
Binary Int32 |  01101000110101111010111101011000 ; Int: 1758965592 ; Date (Int * 1000): 2025-09-27T09:33:12.000Z
Binary Int64 |  0000000001101000000000001101011100000000101011110000000001011000

// For Reference
2017-03-24T04:00:00.000Z
Value: 1490328000000
```

From @FrogTheFrog's example:
```
�6Y
� : 65533 : 11111101 : 1111111111111101
6 : 54 : 00110110 : 0000000000110110
 : 7 : 00000111 : 0000000000000111
Y : 89 : 01011001 : 0000000001011001
Character Codes | Sum: 65683 ; Concat: 6553354789
Binary Int32 |  11111101001101100000011101011001 ; Int: 4248176473 ; Date (Int * 1000): 2104-08-14T17:01:13.000Z
Binary Int64 |  1111111111111101000000000011011000000000000001110000000001011001
```

Also, my previous understanding of booleans appears to be incorrect. Booleans appear to be special cases of numbers, since they share a type indicator of `\u0002` and are represented by four characters as the value (without a value termination character, unlike strings).

**Comment by FrogTheFrog** [comment]
@tirish Nothing makes sense because you think in Big Endian :D. I believe you should rename 3 constants you have from `Object`, `String` and `Boolean` to `v0`, `v1` and `v2` and add comments about what they could be.

I realized that `v2` is in little endian, because value true is written as `\01\00\00\00`, which made no sense. In other words, here is how you should handle `v2`:

```
function fromV2(data) {
    var buffer = Buffer.from(data);
    return buffer.readInt32LE();
}

function toV2(data) {
    var buffer = Buffer.alloc(4);
    buffer.writeInt32LE(data);
    return buffer.toString();
}

//For Date

var data = Buffer.from([104, 215, 175, 88]).toString(); //Ech, just couldn't enter string normally...
var parsedData = fromV2(data);

//Problem here is how to determine that it is a date?
//Specify field name?
//Edit: actually, it should be up to user, not the parser.

console.log(new Date(parsedData * 1000));

//Convert from number

var cantThinkOfAName = toV2(parsedData);
console.log(data, parsedData, cantThinkOfAName, data === cantThinkOfAName);

//For boolean

data = '\01\00\00\00';
parsedData = fromV2(data);
cantThinkOfAName = toV2(parsedData);
console.log(data, parsedData, cantThinkOfAName, data === cantThinkOfAName);
```

**Comment by tirish** [comment]
@FrogTheFrog check out https://github.com/tirish/steam-shortcut-editor/pull/8

**Comment by tirish** [comment]
@FrogTheFrog @SkyNinja 
I have pushed a fix, version 2.2.0 of the package. See updated README for how to use the new parser options to handle date properties. 
TLDR:
```javascript
shortcut.parse(input, { 
  autoConvertBooleans: true, 
  autoConvertArrays: true, 
  dateProperties: ['LastPlayTime'] 
})
```

Let me know if you have any other issues or suggestions.

**Comment by grgohr** [comment]
hello, it seems that dates format has changed again as even with that fix and the proper dates format in the JSON input, Steam considers the non steam games never played:

example:
the parser properly get the date
 {
      appid: -2020228708,
      appname: 'Genshin Impact ',
      exe: '...\\Genshin Impact\\launcher.exe',
      StartDir: '...\\Genshin Impact',
      icon: '',
      ShortcutPath: '',
      LaunchOptions: '',
      IsHidden: false,
      AllowDesktopConfig: true,
      AllowOverlay: true,
      OpenVR: false,
      Devkit: false,
      DevkitGameID: '',
      DevkitOverrideAppID: false,
      LastPlayTime: '2022-06-28T15:11:14.000Z',
      FlatpakAppID: '',
      tags: []
    },

the writer as well as seen in the generated .vdf
LastPlayTime 2022-06-28T15:11:14.000Z 

but still when using it in steam, only steam games keep their actual LastPlayTime, all non steam games become "never played"

**Comment by parkerlreed** [comment]
Also appid seems to be displaying wrong
